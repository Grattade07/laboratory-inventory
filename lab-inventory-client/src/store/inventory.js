import {createSlice, current} from "@reduxjs/toolkit"

/* state slice for the application */
export const inventorySlice = createSlice({
    name: "inventory",

    /* initial set states */
    initialState: {
        inventoryUpdateCount: 0,
        inventoryList: [],
        displayedInventoryList: [],
        userJWToken: "",
        isAdmin: false,
        isLoggedIn: false
    },

    /* reducers to change state */
    reducers: {
        /* increments the update count */
        updateInventoryCount: (state) => {
            state.inventoryUpdateCount += 1
        },
        /* updates the inventory list */
        updateInventoryList: (state, action) => {
            state.inventoryList = action.payload
        },
        /* updates to the default inventory list to be shown in the table */
        displayList: (state) => {
            state.displayedInventoryList = state.inventoryList
        },
        /* makes the displayed table sorted by item name */
        displayItemSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.name > p2.name) ? 1 : (p1.name < p2.name) ? -1 : 0)
        },
        /* makes the displayed table sorted by item location */
        displayLocationSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.location > p2.location) ? 1 : (p1.location < p2.location) ? -1 : 0)
        },
        /* makes the displayed table sorted by item batch */
        displayBatchSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.batch > p2.batch) ? 1 : (p1.batch < p2.batch) ? -1 : 0)
        },
        /* makes the displayed table sorted by item quantity */
        displayQuantitySortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.quantity > p2.quantity) ? 1 : (p1.quantity < p2.quantity) ? -1 : 0)
        },
        /* makes the displayed table sorted by item expiry */
        displayExpirySortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.expiryDate > p2.expiryDate) ? 1 : (p1.expiryDate < p2.expiryDate) ? -1 : 0)
        },
        /* updates the stored JWToken */
        addJWToken: (state, action) => {
            state.userJWToken = action.payload
        },
        /* removes the stored JWToken */
        removeJWToken: (state) => {
            state.userJWToken = ""
        },
        /* changes the displayed table to what was search in the search bar */
        displayFilteredList: (state, action) => {
            let filterString = action.payload
            /* current function gets the current state of the inventoryList state. Found this function from url: https://redux-toolkit.js.org/api/other-exports#current */
            let inventoryList = current(state.inventoryList)
            let filteredList = function(string) {
                let filteredList = []

                for (let i = 0; i < inventoryList.length; i++) {
                    if(inventoryList[i].name.toLowerCase().includes(string)) {
                        filteredList.push(inventoryList[i])
                    }
                }

                return filteredList
            }

            state.displayedInventoryList = filteredList(filterString)
        },
        /* resets the displayed list to the original data */
        resetDisplayList: (state) => {
            state.displayedInventoryList = state.inventoryList
        },
        /* updates if user is admin */
        setIsAdmin: (state, action) => {
            state.isAdmin = action.payload
        },
        /* updates if user is logged in */
        setIsLoggedIn: (state, action) => {
            state.isLoggedIn = action.payload
        }
    }
})

export const {
    updateInventoryCount, 
    updateInventoryList, 
    displayList, 
    displayItemSortedList, 
    displayLocationSortedList, 
    displayBatchSortedList, 
    displayQuantitySortedList, 
    displayExpirySortedList, 
    addJWToken, 
    removeJWToken, 
    displayFilteredList, 
    resetDisplayList,
    setIsAdmin,
    setIsLoggedIn
} = inventorySlice.actions

export default inventorySlice.reducer