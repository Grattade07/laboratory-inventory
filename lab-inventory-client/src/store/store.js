import { configureStore } from "@reduxjs/toolkit";
import inventoryReducer from "./inventory"

/* store for application */
export default configureStore({
    reducer: {
        inventory : inventoryReducer
    }
})