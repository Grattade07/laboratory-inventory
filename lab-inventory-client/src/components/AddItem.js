import Form from "react-bootstrap/Form"
import Button from "react-bootstrap/Button";
import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { updateInventoryCount } from "../store/inventory";

/* component used to add new items to the inventory database */
function AddItem () {
    const [itemName, handleItemName] = useState("")
    const [location, handleLocation] = useState("")
    const [batch, handleBatch] = useState("")
    const [quantity, handleQuantity] = useState("")
    const [expiryDate, handleExpiryDate] = useState("")
    const [purchaseUrl, handlePurchaseUrl] = useState("")

    const dispatch = useDispatch()

    const token = useSelector((state) => state.inventory.userJWToken)

    const isAdmin = useSelector((state) => state.inventory.isAdmin)

    /* function sends post request to add new item to database */
    function postNewItem(url = "", data = {}) {
        return fetch(url, {
            method: "POST",
            headers: {
                "Content-Type" : "application/json",
                "token" : token
            },
            body: JSON.stringify(data)
        })
        .then(res => res.json())
        .then(dispatch(updateInventoryCount()))
    }

    /* function resets all input values */
    function resetInput() {
        handleItemName("")
        handleLocation("")
        handleBatch("")
        handleQuantity("")
        handleExpiryDate("")
        handlePurchaseUrl("")
    }

    return (
        <section id="add-item-container">

            <div>
                <h4 className="underline">Add New Item</h4>
            </div>
            
            <Form id="add-item-form" >
                {/* input for item name */}
                <Form.Group className="underline">
                    <Form.Label>Item Name:</Form.Label>
                    <Form.Control data-testid="item-add-name" name="name" type="text" placeholder="Example: Ethanol" onChange={(e) => handleItemName(e.target.value)} value={itemName}/>
                </Form.Group>
                {/* input for item location */}
                <Form.Group className="underline">
                    <Form.Label>Location:</Form.Label>
                    <Form.Control data-testid="item-add-location"  name="location" type="text" placeholder="Example: Cupboard" onChange={(e) => handleLocation(e.target.value)} value={location}/>
                </Form.Group>
                {/* input for batch */}
                <Form.Group className="underline">
                    <Form.Label>Batch No.:</Form.Label>
                    <Form.Control data-testid="item-add-batch" name="batch" type="text" placeholder="Example: #1EF43T" onChange={(e) => handleBatch(e.target.value)} value={batch}/>
                </Form.Group>
                {/* input for quantity */}
                <Form.Group className="underline">
                    <Form.Label>Quantity:</Form.Label>
                    <Form.Control data-testid="item-add-quantity"  name="quantity" type="text" placeholder="Example: 53" onChange={(e) => handleQuantity(e.target.value)} value={quantity}/>
                </Form.Group>
                {/* input for expiry date */}
                <Form.Group className="underline">
                    <Form.Label>Expiry Date:</Form.Label>
                    <Form.Control data-testid="item-add-expiry"  name="expiry" type="date" onChange={(e) => handleExpiryDate(e.target.value)} value={expiryDate}/>
                </Form.Group>
                {/* input for url to purchase item */}
                <Form.Group className="underline">
                    <Form.Label>Purchase URL:</Form.Label>
                    <Form.Control data-testid="item-add-url"  name="url" type="text" placeholder="Example: www.amazon.co.uk" onChange={(e) => handlePurchaseUrl(e.target.value)} value={purchaseUrl}/>
                </Form.Group>

                {/* button sends request to add new item. Is disabled if user is not an admin */}
                <Button variant="secondary" id="add-item-submit"onClick={() => {
                    postNewItem("/inventory/createItem", {
                    name: itemName,
                    location: location,
                    batch: batch,
                    quantity: quantity,
                    expiryDate: expiryDate,
                    url: purchaseUrl
                    })
                    resetInput()
                }} disabled={isAdmin ? false : true}>Submit</Button>
            </Form>
        </section>
    )
}

export default AddItem