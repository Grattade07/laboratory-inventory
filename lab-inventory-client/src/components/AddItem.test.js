import renderer from "react-test-renderer"
import AddItem from "./AddItem"
import { Provider } from 'react-redux';
import store from '../store/store';
import {render} from "@testing-library/react"

/* tests that the add item component renders correctly. Found how to do from url: https://www.digitalocean.com/community/tutorials/how-to-write-snapshot-tests-for-react-components-with-jest  */
it("renders correctly", () => {
    const tree = renderer
        .create(<Provider store={store}>
                    <AddItem />
                </Provider>
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})

/* unit test makes sure that the inputs to add an item render with no value. Found how to do this from url: https://levelup.gitconnected.com/how-to-write-unit-tests-with-react-testing-library-d9624fd2b707 */
describe("AddItem component", () => {
    const mockChangeValue = jest.fn()
    const stubbedSearchValue = {
        name: "",
        location: "",
        batch: "",
        quantity: "",
        expiry: "",
        url: ""
    }

    it("Shows all input fields with empy values", () => {
        const {getByTestId} = render(
            <Provider store={store}>
                <AddItem />
            </Provider>
        )

        expect(getByTestId("item-add-name").value).toBe("")
        expect(getByTestId("item-add-location").value).toBe("")
        expect(getByTestId("item-add-batch").value).toBe("")
        expect(getByTestId("item-add-quantity").value).toBe("")
        expect(getByTestId("item-add-expiry").value).toBe("")
        expect(getByTestId("item-add-url").value).toBe("")
    })
})