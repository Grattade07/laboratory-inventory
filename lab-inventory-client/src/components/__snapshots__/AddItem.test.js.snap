// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  id="add-item-container"
>
  <div>
    <h4
      className="underline"
    >
      Add New Item
    </h4>
  </div>
  <form
    className=""
    id="add-item-form"
  >
    <div
      className="underline"
    >
      <label
        className="form-label"
      >
        Item Name:
      </label>
      <input
        className="form-control"
        data-testid="item-add-name"
        name="name"
        onChange={[Function]}
        placeholder="Example: Ethanol"
        type="text"
        value=""
      />
    </div>
    <div
      className="underline"
    >
      <label
        className="form-label"
      >
        Location:
      </label>
      <input
        className="form-control"
        data-testid="item-add-location"
        name="location"
        onChange={[Function]}
        placeholder="Example: Cupboard"
        type="text"
        value=""
      />
    </div>
    <div
      className="underline"
    >
      <label
        className="form-label"
      >
        Batch No.:
      </label>
      <input
        className="form-control"
        data-testid="item-add-batch"
        name="batch"
        onChange={[Function]}
        placeholder="Example: #1EF43T"
        type="text"
        value=""
      />
    </div>
    <div
      className="underline"
    >
      <label
        className="form-label"
      >
        Quantity:
      </label>
      <input
        className="form-control"
        data-testid="item-add-quantity"
        name="quantity"
        onChange={[Function]}
        placeholder="Example: 53"
        type="text"
        value=""
      />
    </div>
    <div
      className="underline"
    >
      <label
        className="form-label"
      >
        Expiry Date:
      </label>
      <input
        className="form-control"
        data-testid="item-add-expiry"
        name="expiry"
        onChange={[Function]}
        type="date"
        value=""
      />
    </div>
    <div
      className="underline"
    >
      <label
        className="form-label"
      >
        Purchase URL:
      </label>
      <input
        className="form-control"
        data-testid="item-add-url"
        name="url"
        onChange={[Function]}
        placeholder="Example: www.amazon.co.uk"
        type="text"
        value=""
      />
    </div>
    <button
      className="btn btn-secondary"
      disabled={true}
      id="add-item-submit"
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </form>
</section>
`;
